### PR TITLE
Bugfix: deadlock when domain failover

### PR DESCRIPTION
**Pull request by wxing1292**
* fix deadlock issue in timer queue processor failover
* fix domain cache copy without locking issue

### CODE REVIEW

File: service/history/timerQueueProcessor.go [hunk 1]
```diff
@@ -143,7 +143,9 @@ func (t *timerQueueProcessorImpl) FailoverDomain(domainID string) {
 	failoverTimerProcessor := newTimerQueueFailoverProcessor(t.shard, t.historyService, domainID,
 		standbyClusterName, minLevel, maxLevel, t.matchingClient, t.logger)
 	failoverTimerProcessor.Start()
-	failoverTimerProcessor.timerQueueProcessorBase.readAndFanoutTimerTasks()
+	// use a fake timer to trigger db scan
```
Comment (samarabbas) [inline service/history/timerQueueProcessor.go]: 'readAndFanoutTimerTasks' is still called from other places and I thought that is the one which could block.

File: common/cache/domainCache.go [hunk 2]
```diff
@@ -185,13 +188,13 @@ func (c *domainCache) GetAllDomain() map[string]*DomainCacheEntry {
 // afterCallback will be invoked when NOT holding the domain cache lock.
 func (c *domainCache) RegisterDomainChangeCallback(shard int, initialNotificationVersion int64, beforeCallback CallbackFn, afterCallback CallbackFn) {
 	c.Lock()
-	defer c.Unlock()
-
 	c.beforeCallbacks[shard] = beforeCallback
 	c.afterCallbacks[shard] = afterCallback
+	domainNotificationVersion := c.domainNotificationVersion
+	c.Unlock()
```
Comment (samarabbas) [inline common/cache/domainCache.go]: Looks like both before and after callbacks were called within the lock before this change and now they are called outside of the lock.  Do we even need 2 separate callbacks?

File: service/history/timerQueueProcessor.go (hunk 1 above)
Comment (wxing1292) [reply service/history/timerQueueProcessor.go]: readAndFanoutTimerTasks will be called by an separate go routine, so we should be fine.

File: common/cache/domainCache.go (hunk 2 above)
Comment (wxing1292) [reply common/cache/domainCache.go]: we do not really need a 2 type of callbacks, one before one after, i thought providing 2 type of callbacks can be useful

**Pull request merged**